### ISSUE TITLE
Restructure Learning_role

### ISSUE THREAD

**Issue by kim-mskw** [assume-framework/assume#154]
We encountered some design problems for the policy_update in how the architecture is now. 
Currently, the actor lies in the learning strategy, which is necessary to allow the preloading of already trained actors without the need for the buffer learning_role an policy update. 

To perform a policy update with the actor in that place, the learning role needs access to all units with a learning strategy. That seems weird. 

Instead, we thought about making a learning class similar to the forecaster and make that accessible to all units. This would omit the message sending for writing in the buffer and solve the actor update more straightforwardly. What do you think?

**Comment by kim-mskw** [comment]
The way it is now:
![image](https://github.com/assume-framework/assume/assets/117921871/64cd5dca-2ed0-40de-bffa-d3b09142fd7b)

**Comment by kim-mskw** [comment]
The way Nick and I discussed:
![image](https://github.com/assume-framework/assume/assets/117921871/c77dc02f-a29b-464b-af68-5ac85f555c6d)

**Comment by maurerle** [comment]
For now, we have the learning_role to schedule the actual learning.

In the new approach all learning agents have access to the same LearningClass and the LearningRole is still needed to schedule the actual learning?

The Learning_class then includes everything needed for learning, including the specified algorithm (matd3).
And the Learning_class is known to all learning agents (to predict and play back the experience) as well as the corresponding learning_role (to trigger the updates).

> To perform a policy update with the actor in that place, the learning role needs access to all units with a learning strategy. That seems weird.

I think we still have the problem that, we need to know all learning agents when we initialize the Buffer (with correct size). When would this happen..?

**Comment by nick-harder** [comment]
We can initialize the buffer inside the learning class as well once we are done loading all the agents and know all the numbers. The biggest question is how to trigger the update? Should we still keep a role around just for this or we can make a check inside the run_simulation function which checks the clock and if it is time to update calls the learning_role.update()?

**Comment by maurerle** [comment]
If there is something to do in given intervals, it should be a Mango Agent with a scheduled task.

Changing this would not fit to our event-based simulation approach

**Comment by kim-mskw** [comment]
okay, so then the approach with having the role for the scheduling and the class for giving to the units such as the forecaster seems like a valid solution. I think we cannot avoid this ugly loop that tells us how many learning agents there are.

Or would it be easier to give each unit the learning_role directly?

**Comment by kim-mskw** [comment]
Then we have the problem again that the buffer is created all the time don't we?

**Comment by maurerle** [comment]
Thanks for the images, which helped with a deeper understanding of what is needed for the learning.

In #157 I added the critics to the LearningRole.
The LearningRole now has direct access to the LearningStrategies and can set the values to the target_actor and actor accordingly.
So its a derivation of the proposed solution - the LearningRole has access to the Buffer and Critics and schedules the `policy_update` - so it also functions as the proposed learning_class.

>  3. This ID should be not the unit id but maybe with something extra so we can differentiate in the future in case units might have several startegies that are learning. so maybe like united_market

True, this is currently missing in #157 


If we do not send messages, we have no option to run the learning in a different process.
This is not a problem, as we are always waiting for the learningrole anyway.
But one can see that self.buffer stays empty, if we set `same_process = False` for the LearningRole, as we can not set the value of a different process.
